design/design-doc-developer-experience-tooling.md

```diff
@@ -0,0 +1,1146 @@
+Crossplane package source trees. As described below, the build tooling
+constructs a `crossplane.yaml` for each package it produces based on the
+contents of the project, including the `project.yaml`.
+
+When a user runs `crossplane project build`, four Crossplane packages will be
+produced: a Configuration and three Functions. The Configuration will include
+automatically generated dependencies on the functions, as well as any additional
+dependencies specified in `project.yaml`. Automated tests can be executed with
+`crossplane project test run`, and the project can be installed on a local
+development control plane for manual testing with `crossplane project
+run`. Running `crossplane project push` will push all four packages to a
+registry.
+
+### The Project File
+
+The project configuration file (`project.yaml` by default, overridable with a
+CLI argument) looks like this:
+
+```yaml
+apiVersion: dev.crossplane.io/v1alpha1
+kind: Project
+metadata:
+  name: my-platform
+spec:
+  # These optional fields are converted to Configuration annotations.
+  maintainer: "Platform Team <platform@example.com>"
+  source: github.com/examplecom/my-platform
+  license: Apache-2.0
+  description: An example configuration using functions.
+  readme: This is just an example.
+
+  # OCI repository where the project will be pushed. This is used as part of the
+  # build process to construct dependencies on the embedded functions.
+  repository: ghcr.io/examplecom/my-platform
+
+  # Crossplane version constraints (optional).
+  crossplane:
+    version: ">=v1.17.0-0"
+
+  # External dependencies (optional).
+  dependsOn:
+    - provider: xpkg.crossplane.io/crossplane-contrib/provider-nop
+      version: ">=v0.2.1"
+    - function: xpkg.crossplane.io/crossplane-contrib/function-auto-ready
+      version: ">=v0.2.1"
+
+  # Dependencies that are used only for authoring functions, but do not need to
+  # be installed as dependencies of the configuration (optional).
+  apiDependencies:
+    - type: k8s
+      k8s:
+        version: v1.33.0
+    - type: crd
+      git:
+        repository: github.com/kubernetes-sigs/cluster-api
+        ref: v1.11.3
+        path: config/crd/bases
+
+  # Where the build tooling should look for various parts of the configuration,
+  # relative to the location of the metadata file. (optional).
+  paths:
+    apis: apis
+    examples: examples
+    functions: functions
+    operations: operations
+    tests: tests
+
+  # Architectures for which to build functions (optional).
+  architectures:
+    - amd64
+    - arm64
+
+  # Optional image configs to rewrite package locations during development, for
+  # example to enable use of the DevEx tools in network restricted environments.
+  imageConfigs:
+    - matchImages:
+        - type: prefix
+          prefix: xpkg.crossplane.io/crossplane-contrib
+      rewriteImage:
+        prefix: internal-registry.example.com/mirror/crossplane-contrib
```

Comment:
What's the benefit of reproducing (a subset of?) the ImageConfig schema here inline? Would taking a dir of ImageConfigs work just a well?